### ISSUE TITLE
Telemetry items not immediately available for lua widgets.

### ISSUE THREAD

**Issue by noamkrief** [FrSkyRC/ETHOS-Feedback-Community#4482]
Hello.

We feel that telemetry is not fully loaded.
I wish that a widget would wait to even load till the full model has loaded.

in wakeup() {
			if (system.getSource("RPM"):state() == true) then
				print('true')
			else
				print('false')
			end
}

You will see the first few loops print false.
Then they print true forever.

You can replicate this on the simulator.exe

Thank you :)

**Comment by robthomson** [comment]
Is this actually an issue?  - surely just a case of coding your lua to look for sensors until initialised? 

Not found it to ever be an issue to me and I have made quite a few widgets. 

Simple example here :

https://github.com/robthomson/RF2GOV/blob/main/scripts/rf2gov/rf2gov.lua

**Comment by Alex00101** [comment]
Telemetry takes time to initialise after receiver is connected. Which may be minutes after the model is loaded and widget started. You just need to handle it properly. 1st thing is not to call getSource in wakeup loop. Then you can use state() or globally telemetry active state.

**Comment by bsongis-frsky** [comment]
As @Alex00101 explains, `state()` will return true only after some value has been received on this sensor. 

He is always right in telling you to call `system.getSource("RPM")` which is very time consuming!

**Comment by noamkrief** [comment]
@bsongis-frsky @Alex00101 @robthomson 
Sorry for the confusion.
This was just the proof that when I getSource in create() the status is false.
Also that :value() in wakeup() in the first 0.5 seconds is nil, which can lead to errors.

**Comment by bsongis** [comment]
`value()` will return nil until something is received and it is correct!

**Comment by noamkrief** [comment]
@bsongis 
This is not what I'm experiencing.
I just turned on the radio with the RX off.
And I'm painting value of RPM and getting 0.0

lcd.drawText(1, 1, 'hello'.. system.getSource({name="RPM"}):value()) --debug

**Comment by robthomson** [comment]
Ping me a copy of your code if you fancy and I will look at it.. But all up issue I believe is that you are not catering well for how these things work.

The behaviour is 100% expected and correct.

**Comment by noamkrief** [comment]
@robthomson 
I'm helping out a friend to clean his widget. So I'm not at liberty to share the code.
I can say that we are connecting to all telemetry items needed in create()
And only getting their :value() in wakeup at 3 times per second.

I'm slowly helping him move everything out of paint. Because paint only needs to paint. It shouldn't do any other stuff in my opinion.


But what @bsongis  is not correct I think "value() will return nil until something is received and it is correct!"
Because I get 0.0 when I do:
lcd.drawText(1, 1, 'hello'.. system.getSource({name="RPM"}):value()) --debug
When the TX is completely off.

So I'm seeking some clarification...

**Comment by robthomson** [comment]
I assume you mean when the rx is turned off? 

If the rx is turned off after initialising I would normally expect to see the last value being sent. 

That being said I would also normally in my code check the link is up before even looking for a sensor. 

Can you apply a stripped down version of the code? 

Or reach me on discord so we can share code privately?  I am on the frsky and rotorflight and inav discords.

**Comment by bsongis-frsky** [comment]
This code already raises an exception when the "RPM" sensor doesn't exist.
It will also raise an exception in 1.5.16 because `value()` on a sensor which has received no value since startup will also return nil (not 0 as of today).

```lua
lcd.drawText(1, 1, 'hello'.. system.getSource({name="RPM"}):value())
```

**Comment by robthomson** [comment]
I ram over this with @noamkrief last night via separate private channel.

Looks like he is all sorted now and knows best way forward.

**Comment by i3dm** [comment]
> This code already raises an exception when the "RPM" sensor doesn't exist. It will also raise an exception in 1.5.16 because `value()` on a sensor which has received no value since startup will also return nil (not 0 as of today).
> 
> ```lua
> lcd.drawText(1, 1, 'hello'.. system.getSource({name="RPM"}):value())
> ```

why is this changed? 0 is much easier to handle than nil. and will make lua development easier without getting value exceptions.

**Comment by Alex00101** [comment]
> > This code already raises an exception when the "RPM" sensor doesn't exist. It will also raise an exception in 1.5.16 because `value()` on a sensor which has received no value since startup will also return nil (not 0 as of today).
> > ```lua
> > lcd.drawText(1, 1, 'hello'.. system.getSource({name="RPM"}):value())
> > ```
> 
> why is this changed? 0 is much easier to handle than nil. and will make lua development easier without getting value exceptions.

I can only guess.... assuming 0 dould be a valid telemetry value , while nil indicates there's no value at all.

**Comment by i3dm** [comment]
my question is for @bsongis-frsky

**Comment by i3dm** [comment]
> This code already raises an exception when the "RPM" sensor doesn't exist. It will also raise an exception in 1.5.16 because `value()` on a sensor which has received no value since startup will also return nil (not 0 as of today).
> 
> ```lua
> lcd.drawText(1, 1, 'hello'.. system.getSource({name="RPM"}):value())
> ```

this is a terrible feature that will break many user widgets. it sure broke all my zavionix widgets.
why is this feature needed? we were working fine without it.

**Comment by robthomsonb4b** [comment]
surely you should be doing this:

```
local src = system.getSource({name="RPM"})
local value
if src ~= nil then
  value = src:value()
  if value == nil then
      value = 'something'
  end
else
   value = 'something'
 end

lcd.drawText(1, 1, 'hello'.. value)
```
Any other approach essentially is not checking your var is valid.

**Comment by i3dm** [comment]
> surely you should be doing this:
> 
> ```
> local src = system.getSource({name="RPM"})
> local value
> if src ~= nil then
>   value = src:value()
>   if value == nil then
>       value = 'something'
>   end
> else
>    value = 'something'
>  end
> 
> lcd.drawText(1, 1, 'hello'.. value)
> ```
> 
> Any other approach essentially is not checking your var is valid.

this would mean running getSource many times, which will load up the widget.
i dont think this is a very good solution.

even if you stop using getSource once the sensor is valid, it may still become invalid later and crash the widget - so how will we know?

**Comment by robthomson** [comment]
You can cache the source.

I do this quite effectively in the rotorflight lua scripts.  Simple fast and effective.

By memory there is an example in here

https://github.com/rotorflight/rotorflight-lua-ethos-suite/blob/master/scripts/rfsuite/tasks/telemetry/telemetry.lua

In this case handling correct sensor acquisition weather crsf or fport

You can also periodically poll to check if valid or invalid.

In rotorflight I actually use a background task that runs as a service.

This then handles all calls between telemetry, fbl, msp etc..   Works rather well.

**Comment by i3dm** [comment]
> You can cache the source.
> 
> I do this quite effectively in the rotorflight lua scripts. Simple fast and effective.
> 
> By memory there is an example in here
> 
> https://github.com/rotorflight/rotorflight-lua-ethos-suite/blob/master/scripts/rfsuite/tasks/telemetry/telemetry.lua
> 
> In this case handling correct sensor acquisition weather crsf or fport
> 
> You can also periodically poll to check if valid or invalid.
> 
> In rotorflight I actually use a background task that runs as a service.
> 
> This then handles all calls between telemetry, fbl, msp etc.. Works rather well.

this is definitely above my knowledge.
and why go through all this trouble to ruin something that worked perfectly well before? what was so bad with a default value of 0 for an inactive sensor? in the old method anybody could poll sensor.state() !
this makes absolutely no sense to me @bsongis-frsky it ruined a perfectly good mechnism.
i would like to request going back to the old way pre 1.5.18. it was much better in my opinion.

**Comment by bsongis** [comment]
I don't think that return 0 is a good idea when nothing has been received. 
There were also sensors calculated from an interpolation table or which had an offset applied. In that case the value was not even 0 but the value of the offset, or the interpolation table at position 0. Which was indeed very wrong. So return nil seems perfectly fine when nothing can be returned.

**Comment by i3dm** [comment]
> I don't think that return 0 is a good idea when nothing has been received. There were also sensors calculated from an interpolation table or which had an offset applied. In that case the value was not even 0 but the value of the offset, or the interpolation table at position 0. Which was indeed very wrong. So return nil seems perfectly fine when nothing can be returned.

this yields a huge effort for us mortal developers in always checking if a sensor source and its vaule isnt nil, and if theres the smallest bug the widget will crash completely.
sorry this makes no sense. there must be a better way,

**Comment by noamkrief** [comment]
This is brutal :) Just brutal.
Please kindly remember that many (if not all) widget developers do this for the benefit of others and share wonderful widgets for free with lots of hard work and effort.
Some may be professional programmers, some may not be. 
But these free widgets are benefiting frsky users worldwide and frsky itself as a company.

I used to be able to do inside paint() the following:

lcd.drawText(190, 108, "("..vfas:value() .."V)")

or 

if bat1percent:value() < 50 or bat2percent:value() < 50 then


Now, we have to do a lot of extra work and coding to protect what can be countless lines like this.

I wouldn't have fixed this "bug" but rather made a note on the lua docs about its behavior.

Oh well - time to get to work :)
Cheers

**Comment by robthomson** [comment]
I kind of think despite the fact it is brutal - it's still the right thing to do.

Right now ethos does not have that many lua devs.  It's on the cusp of growing rapidly.

Would you rather fix a problem while there is low impact?  Or leave it for years where it becomes a bigger issue.

Totally.. A pita.  But prob still the right choice.

**Comment by i3dm** [comment]
It's not a question of what's technically right. But the impact on people doing 3 years of work for free.

I personally found yesterday that I now have to do >100 hours of work for practically nothing to get the my widgets to run propey again just because it is "technically better" despite having the old way working fine for years and still allowing user to check sensor. State().

I'm sorry but Noam is right. Since this community is based on volunteers, it makes no sense implementing such a wide spread change that kills all widgets.

Just think about the end users who had Zavionix widgets working great for 3 years and now all are broken! This is such a bad move for Frsky...
The end users know nothing about Lua and you are literally killing their perfectly working system for no gain what so ever to the user.  

It is crazy not to consider the consequences of this move.

**Comment by Alex00101** [comment]
I agree that the change could have been better documented/announced, but I think the above is taking it too far.
1st, the change is well aligned with how telemetry works and its state transition model. I have been through the situations where 0 is the right return value from the sensor. I agree with Rob who posted that it is better to change now and absorb the impact rather than keep the ambiguity for the sake of backward compatibility. 1.5.18 for example fixes the return value of curve point where the documentation tells me its int while in fact (before 1.5.18) it was returning a table. I had to change my code, so be it, the reality is backwards compatibility is not always possible.
2nd, whatever the value is it does not replace safe coding practices - i.e. checking something is not nil before referring to it, etc. Specifically Ethos provides a way of checking sensor and telemetry state for various scenarios. If this is not currently used, then there is a chance to improve the reliability.

**Comment by robthomson** [comment]
Simple fact is that the 1.5 series has been going through a large number of functional changes.

Many new features have been added.  Many lua options have been deprecated.

When I started my rotorflight lua 6 months ago - it was not possible to achieve the leval of integration required without major updates to lua.  Some of them where breaking changes.

But what we do have now is an extremely capable platform that is far better positioned for the future.

One thing I 100% do in my code is a version check on ethos as it fires up.  While ethos is rapidly updating it is really helpful to know if you are or are not supported.

Could it have been better communicated.. Probably yes. 

Maybe what is needed is a developer mailing list.   Or maybe a more active developer frsky discord group (currently very under utilised)

I think I realised that change back at 1.5.16..   Took maybe 15 mins to resolve as all my sensor acquisition is already abstracted by a wrapper function.

Maybe for you guys who have difficulty changing your code....

Make a simple wrapper function that simply returns the data in the way you expect.  That way you only have to change a function name.. And do the logic once?
